packages/task/src/browser/task-service.ts

```diff
@@ -238,18 +290,58 @@ export class TaskService implements TaskConfigurationClient {
      */
     async run(source: string, taskLabel: string): Promise<void> {
         let task = await this.getProvidedTask(source, taskLabel);
-        if (!task) {
+        const matchers: (string | ProblemMatcherContribution)[] = [];
+        if (!task) { // if a provided task cannot be found, search from tasks.json
             task = this.taskConfigurations.getTask(source, taskLabel);
             if (!task) {
                 this.logger.error(`Can't get task launch configuration for label: ${taskLabel}`);
                 return;
+            } else if (task.problemMatcher) {
+                if (Array.isArray(task.problemMatcher)) {
+                    matchers.push(...task.problemMatcher);
+                } else {
+                    matchers.push(task.problemMatcher);
+                }
             }
+        } else { // if a provided task is found, check if it is customized in tasks.json
+            const taskType = task.taskType || task.type;
+            const customizations = this.taskConfigurations.getTaskCustomizations(taskType);
+            const matcherContributions = this.getProblemMatchers(task, customizations);
+            matchers.push(...matcherContributions);
         }
-
-        this.runTask(task);
+        await this.problemMatcherRegistry.onReady();
+        const resolvedMatchers: ProblemMatcher[] = [];
+        // resolve matchers before passing them to the server
+        for (const matcher of matchers) {
+            let resolvedMatcher: ProblemMatcher | undefined;
+            if (typeof matcher === 'string') {
+                resolvedMatcher = this.problemMatcherRegistry.get(matcher);
+            } else {
+                resolvedMatcher = await this.problemMatcherRegistry.getProblemMatcherFromContribution(matcher);
+            }
+            if (resolvedMatcher) {
+                const scope = task._scope || task._source;
+                if (resolvedMatcher.filePrefix && scope) {
+                    const options = { context: new URI(scope).withScheme('file') };
+                    const cwd = await this.variableResolverService.resolve(task.cwd.toString(), options);
+                    Object.assign(options, {
```

Comment:
Should not we register `cwd` variable instead in the variable registry generally? And remove new APIs to the resolver.

It's generally always available in VS Code: https://github.com/microsoft/vscode/blob/466c132f5b4362b3be485a1f7a624b3a7316ddb8/src/vs/workbench/services/configurationResolver/common/variableResolver.ts#L210-L211

Reply:
like i mentioned in another comment,  variables like ${cwd} could possibly be different from one task to another, e.g.,
```js
[
{ type: 'shell', label: 'task 1', cwd: '/home/user/theia', ... },
{ type: 'shell', label: 'task 2', cwd: '/home/user', ... }
]
```

yes, ${cwd} is generally available in vsCode. but i found the variableResolver in vsCode allows users to pass a uri into the function:
https://github.com/microsoft/vscode/blob/466c132f5b4362b3be485a1f7a624b3a7316ddb8/src/vs/workbench/services/configurationResolver/common/variableResolver.ts#L93
and therefore every time ${cwd} gets resolved, the resolver returns the uri passed through. 

do we want to do the same thing?

Reply:
Don't we pass context URI from which we can derive a folder URI already?

Reply:
OK I updated the code to match vscode from the functionality stand point.